lib/response.js

```diff
@@ -16,31 +20,57 @@ class Response {
     this.server = server
     this.socket = socket
     this.statusCode = 200
+    this.statusMessage = undefined
     this.headerSent = false
 
     this._headers = headers
     this._headersLength = 0
+    this._headersMap = new Map()
     this._keepAlive = true
     this._chunked = true
     this._reuseChunkHeader = server._reuseChunkHeader
     this._reuseChunk = server._reuseChunk
   }
 
-  setHeader (name, value) {
-    const header = name + SEP + value + EOL
+  hasHeader (name) {
+    return this._headersMap.has(name.toLowerCase())
+  }
 
-    // slow path but very unlikely (a *lot* of headers)
-    if (this._headersLength + header.length > 65534) {
-      this._headers = Buffer.concat([this._headers, Buffer.allocUnsafe(65536)])
-    }
+  getHeader (name) {
+    const entry = this._headersMap.get(name.toLowerCase())
+    if (!entry) return
+    return entry[1]
+  }
 
-    this._headers.asciiWrite(header, this._headersLength, header.length)
-    this._headersLength += header.length
+  getHeaders () {
+    const headers = {}
+    this._headersMap.forEach(entry => {
+      headers[entry[0]] = entry[1]
+    })
+    return headers
+  }
 
+  setHeader (name, value) {
+    this._headersMap.set(name.toLowerCase(), [name, value])
     if (CONTENT_LENGTH.test(name)) this._chunked = false
     else if (CONNECTION.test(name)) this._keepAlive = false
   }
 
+  writeHead (statusCode, statusMessage, headers = {}) {
+    this.statusCode = statusCode
+
+    if (typeof statusMessage === 'string') {
+      this.statusMessage = statusMessage
+    } else {
+      this.statusMessage = httpStatusMessage[this.statusCode]
+      headers = statusMessage
+    }
```

Comment:
does this work with writeHead(num) since statusMessage will be undefined?

Reply:
yes, this is not ok. I will fix it